messagix/client.go

```diff
@@ -217,6 +218,7 @@ func (c *Client) Connect() error {
 		return err
 	}
 	ctx, cancel := context.WithCancel(context.TODO())
+	c.ctx = ctx
 	oldCancel := c.stopCurrentConnection.Swap(&cancel)
```

Comment:
I just realized that this context is not cancelled until a new `Connect` call, and that it's not actually per socket connection because this function has a loop that retries socket connections forever (until stopped)

It might actually be better to not mess with contexts at all and instead just close all channels when socket.Connect returns. Basically add a `cancelAllRequests` to `ResponseHandler` which acquires the lock, loops through every request and packet channel and closes them. The places that consume the channels will then get a `nil`, which can be interpreted as the request being cancelled due to disconnection

Reply:
> I just realized that this context is not cancelled until a new Connect call

The context is also cancelled here: 
```go
func (c *Client) Disconnect() {
	if fn := c.stopCurrentConnection.Load(); fn != nil {
		(*fn)()
	}
	c.socket.Disconnect()
}
```

which is called from `User.unlockedDisconnect()` and therefor from `User.FullReconnect()`. 

> and that it's not actually per socket connection because this function has a loop that retries socket connections forever (until stopped)

what do you mean here 🤔? There is only one `Socket` per `Client`, and therefore only one `ResponseHandler`, isn't that right?

Reply:
Ah yes missed that. Anyway, `Client.Connect` calls `socket.Connect` forever until something tells the client to disconnect (see the loop inside the goroutine a few lines below this). So this will cancel requests during a full reconnect, but not during a reconnect caused by a network error.